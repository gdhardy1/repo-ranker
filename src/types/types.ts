import { Dispatch, SetStateAction } from "react";
import { LazyQueryResult } from "@apollo/client/react/types/types";
export interface OrganizationData {
  organization: Organization;
}

export interface Organization {
  id: string;
  login: string;
  name: string;
  repositories: {
    edges: {
      node: RepositoryNode;
    }[];
    totalCount: number;
    pageInfo: PageInfo;
  };
}

export interface RepositoryNode {
  name: string;
  defaultBranchRef: { name: string };
  id: string;
  stargazers: { totalCount: number };
}
export interface RepositoryData {
  repository: Repository;
}

export interface Repository {
  name: string;
  object: CommitHistory; // git objects filtered by Git rev-parse expression
}

export interface CommitHistory {
  history: {
    totalCount: number;
    edges: {
      node: CommitNode;
    }[];
    pageInfo: PageInfo;
  };
}

export interface CommitNode {
  id: string;
  messageHeadline: string;
  url: string;
}

export interface PageInfo {
  startCursor: string;
  endCursor: string;
  hasNextPage: boolean;
  hasPreviousPage: boolean;
}

export class NullOrganizationData implements OrganizationData {
  organization;

  constructor() {
    this.organization = {
      id: "",
      login: "",
      name: "",
      repositories: {
        edges: [] as { node: RepositoryNode }[],
        totalCount: 0,
        pageInfo: {
          startCursor: "",
          endCursor: "",
          hasNextPage: false,
          hasPreviousPage: false,
        },
      },
    };
  }
}

export class NullRepositoryData implements RepositoryData {
  repository;
  constructor() {
    this.repository = {
      name: "",
      object: {
        history: {
          totalCount: 0,
          edges: [
            {
              node: {
                id: "",
                messageHeadline: "",
                url: "",
              },
            },
          ],
          pageInfo: {
            startCursor: "",
            endCursor: "",
            hasNextPage: false,
            hasPreviousPage: false,
          },
        },
      },
    };
  }
}
export interface OrganizationDataVariables {
  login: string;
  cursor?: string | undefined;
}
export interface RepositoryDataVariables {
  login: string;
  repoName: string;
  branch: string;
  cursor?: string;
}

export type LazyGetter<S, T> = (
  variables: T
) => Promise<LazyQueryResult<S, T> | undefined>;

export type LazyGetterTuple<S, T> = [LazyGetter<S, T>, LazyQueryResult<S, T>];

export type AppContextType = {
  organizationData: OrganizationData;
  setOrganizationData: Dispatch<SetStateAction<OrganizationData>>;
  repositoryData: RepositoryData;
  setRepositoryData: Dispatch<SetStateAction<RepositoryData>>;
  nextCursor: string | undefined;
  setNextCursor: Dispatch<SetStateAction<string | undefined>>;
  notification: Notification;
  setNotification: Dispatch<SetStateAction<Notification>>;
};

export interface Notification {
  message: string;
  title: string;
  show: boolean;
}
