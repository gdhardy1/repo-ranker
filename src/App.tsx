import React, { useState, createContext } from "react";
import "./App.css";
import {
  AppContextType,
  OrganizationData,
  RepositoryData,
  NullOrganizationData,
  NullRepositoryData,
} from "./types/types";
import Search from "./components/Search";
import { Outlet } from "react-router-dom";

export const AppContext = createContext({} as AppContextType);

function App() {
  const [organizationData, setOrganizationData] = useState<OrganizationData>(
    new NullOrganizationData()
  );
  const [repositoryData, setRepositoryData] = useState<RepositoryData>(
    new NullRepositoryData()
  );
  const [startCursor, setStartCursor] = useState("");
  const [nextCursor, setNextCursor] = useState<string | undefined>(undefined);
  const [totalRepoCount, setTotalRepoCount] = useState(0);
  const [login, setLogin] = useState<string>("");
  const context = {
    organizationData,
    setOrganizationData,
    repositoryData,
    setRepositoryData,
    nextCursor,
    setNextCursor,
  };

  return (
    <div className="App">
      <AppContext.Provider value={context}>
        <Search />
        <Outlet />
      </AppContext.Provider>
    </div>
  );
}

export default App;
