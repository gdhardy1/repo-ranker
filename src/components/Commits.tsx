import React, { useContext, useEffect } from "react";

import { useParams } from "react-router-dom";
import { AppContext } from "../App";
import { useRepositoryData } from "../util/hooks/useRepositoryData";

export default function Commits(props: any) {
  const { repositoryData, setRepositoryData } = useContext(AppContext);

  const params = useParams();
  const { branch, repoName, organizationLogin } = params;
  const [getRepositoryData] = useRepositoryData();

  useEffect(() => {
    if (organizationLogin && branch && repoName) {
      const result = getRepositoryData({
        login: organizationLogin,
        repoName,
        branch,
      })
        .then((result) => {
          console.log(result);
          if (result?.data?.repository) {
            setRepositoryData(result.data);
          }
        })
        .catch((err) => {
          console.log(err);
        });
    }
  }, [branch, repoName]);
  return (
    <>
      {repositoryData.repository.object.history.edges.map((edge: any) => {
        const { node: commit } = edge;
        return (
          <li key={commit.id}>
            <a href={commit.url} className="block hover:bg-gray-50">
              <div className="px-4 py-4 sm:px-6">
                <div className="flex items-center justify-between">
                  <p className="truncate text-sm font-medium text-indigo-600">
                    {commit.message}
                  </p>
                  <div className="ml-2 flex flex-shrink-0">
                    <p className="inline-flex rounded-full bg-green-100 px-2 text-xs font-semibold leading-5 text-green-800">
                      View on Github
                    </p>
                  </div>
                </div>
              </div>
            </a>
          </li>
        );
      })}
    </>
  );
}
