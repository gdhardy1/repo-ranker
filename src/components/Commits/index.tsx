import React, { useContext } from "react";
import { AppContext } from "../../App";

export default function Commits(props: any) {
  const { repositoryData } = useContext(AppContext);

  return (
    <>
      {repositoryData.repository.object.history.edges.map((edge: any) => {
        const { node: commit } = edge;
        return (
          <li key={commit.id}>
            <a
              href={commit.url}
              rel="noreferrer noopener"
              target="_blank"
              className="block hover:bg-gray-50"
            >
              <div className="px-4 py-4 sm:px-6">
                <div className="flex items-center justify-between">
                  <p className="truncate text-sm font-medium text-indigo-600">
                    {commit.messageHeadline}
                  </p>
                  <div className="ml-2 flex shrink-0 basis-20">
                    <p className="inline-flex w-16  rounded-full bg-green-100 px-2 text-xs font-semibold leading-5 text-green-800">
                      {commit.oid.slice(0, 7)}
                    </p>
                  </div>
                </div>
              </div>
            </a>
          </li>
        );
      })}
    </>
  );
}
