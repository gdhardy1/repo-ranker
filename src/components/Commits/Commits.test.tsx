import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";
import Commits from ".";
import { AppContext } from "../../App";
import * as mocks from "../../util/test/mocks";
import { AppContextType } from "../../types/types";
import renderer from "react-test-renderer";

describe("<Commit />", () => {
  test("Renders correctly", async () => {
    const mockContext: AppContextType = {
      organizationData: mocks.organizationData,
      repositoryData: mocks.repositoryData,
      nextCursor: "",
      notification: { title: "", message: "", show: false },
      setNextCursor: jest.fn(),
      setOrganizationData: jest.fn(),
      setRepositoryData: jest.fn(),
      setNotification: jest.fn(),
    };

    const tree = renderer
      .create(
        <AppContext.Provider value={mockContext}>
          <Commits />
        </AppContext.Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
