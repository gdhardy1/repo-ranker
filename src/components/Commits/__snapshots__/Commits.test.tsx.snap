// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Commit /> Renders correctly 1`] = `
Array [
  <li>
    <a
      className="block hover:bg-gray-50"
      href="https://github.com/gatsbyjs/gatsby/commit/856386e5fabe7013a36d6e5cc596535383993070"
    >
      <div
        className="px-4 py-4 sm:px-6"
      >
        <div
          className="flex items-center justify-between"
        >
          <p
            className="truncate text-sm font-medium text-indigo-600"
          >
            chore(release): Publish next
          </p>
          <div
            className="ml-2 flex flex-shrink-0"
          >
            <p
              className="inline-flex rounded-full bg-green-100 px-2 text-xs font-semibold leading-5 text-green-800"
            >
              View on Github
            </p>
          </div>
        </div>
      </div>
    </a>
  </li>,
  <li>
    <a
      className="block hover:bg-gray-50"
      href="https://github.com/gatsbyjs/gatsby/commit/ad1148e7d9304b7bec198841f64748242077a1cc"
    >
      <div
        className="px-4 py-4 sm:px-6"
      >
        <div
          className="flex items-center justify-between"
        >
          <p
            className="truncate text-sm font-medium text-indigo-600"
          >
            fix(gatsby): Fix truncation of childNode during static HTML generatio…
          </p>
          <div
            className="ml-2 flex flex-shrink-0"
          >
            <p
              className="inline-flex rounded-full bg-green-100 px-2 text-xs font-semibold leading-5 text-green-800"
            >
              View on Github
            </p>
          </div>
        </div>
      </div>
    </a>
  </li>,
  <li>
    <a
      className="block hover:bg-gray-50"
      href="https://github.com/gatsbyjs/gatsby/commit/45a105b6db6892150da27d71abaf693b8e43fb14"
    >
      <div
        className="px-4 py-4 sm:px-6"
      >
        <div
          className="flex items-center justify-between"
        >
          <p
            className="truncate text-sm font-medium text-indigo-600"
          >
            fix(gatsby): Persist manifest on cached builds (#36693)
          </p>
          <div
            className="ml-2 flex flex-shrink-0"
          >
            <p
              className="inline-flex rounded-full bg-green-100 px-2 text-xs font-semibold leading-5 text-green-800"
            >
              View on Github
            </p>
          </div>
        </div>
      </div>
    </a>
  </li>,
  <li>
    <a
      className="block hover:bg-gray-50"
      href="https://github.com/gatsbyjs/gatsby/commit/2784feef1297d5f33eb6b95f1896a10aa3786d09"
    >
      <div
        className="px-4 py-4 sm:px-6"
      >
        <div
          className="flex items-center justify-between"
        >
          <p
            className="truncate text-sm font-medium text-indigo-600"
          >
            update react peer deps patch (#36707)
          </p>
          <div
            className="ml-2 flex flex-shrink-0"
          >
            <p
              className="inline-flex rounded-full bg-green-100 px-2 text-xs font-semibold leading-5 text-green-800"
            >
              View on Github
            </p>
          </div>
        </div>
      </div>
    </a>
  </li>,
  <li>
    <a
      className="block hover:bg-gray-50"
      href="https://github.com/gatsbyjs/gatsby/commit/a15cdfd382f0fa56e6ff2bf9d41a95bd98062b75"
    >
      <div
        className="px-4 py-4 sm:px-6"
      >
        <div
          className="flex items-center justify-between"
        >
          <p
            className="truncate text-sm font-medium text-indigo-600"
          >
            chore(deps): update starters and examples to ^18.0.21 (#36692)
          </p>
          <div
            className="ml-2 flex flex-shrink-0"
          >
            <p
              className="inline-flex rounded-full bg-green-100 px-2 text-xs font-semibold leading-5 text-green-800"
            >
              View on Github
            </p>
          </div>
        </div>
      </div>
    </a>
  </li>,
  <li>
    <a
      className="block hover:bg-gray-50"
      href="https://github.com/gatsbyjs/gatsby/commit/f300895a768955f34a0181533186fc5277b66b04"
    >
      <div
        className="px-4 py-4 sm:px-6"
      >
        <div
          className="flex items-center justify-between"
        >
          <p
            className="truncate text-sm font-medium text-indigo-600"
          >
            chore(changelogs): update changelogs (#36698)
          </p>
          <div
            className="ml-2 flex flex-shrink-0"
          >
            <p
              className="inline-flex rounded-full bg-green-100 px-2 text-xs font-semibold leading-5 text-green-800"
            >
              View on Github
            </p>
          </div>
        </div>
      </div>
    </a>
  </li>,
  <li>
    <a
      className="block hover:bg-gray-50"
      href="https://github.com/gatsbyjs/gatsby/commit/5117bd14ed229ed15ad62f0cb2136be79ad527ea"
    >
      <div
        className="px-4 py-4 sm:px-6"
      >
        <div
          className="flex items-center justify-between"
        >
          <p
            className="truncate text-sm font-medium text-indigo-600"
          >
            fix(deps): update starters and examples - gatsby (#36706)
          </p>
          <div
            className="ml-2 flex flex-shrink-0"
          >
            <p
              className="inline-flex rounded-full bg-green-100 px-2 text-xs font-semibold leading-5 text-green-800"
            >
              View on Github
            </p>
          </div>
        </div>
      </div>
    </a>
  </li>,
  <li>
    <a
      className="block hover:bg-gray-50"
      href="https://github.com/gatsbyjs/gatsby/commit/57e3363e6fd700e298c6fd94265f8e53864fab69"
    >
      <div
        className="px-4 py-4 sm:px-6"
      >
        <div
          className="flex items-center justify-between"
        >
          <p
            className="truncate text-sm font-medium text-indigo-600"
          >
            chore(docs): add showcase link to 4.24 release notes (#36705)
          </p>
          <div
            className="ml-2 flex flex-shrink-0"
          >
            <p
              className="inline-flex rounded-full bg-green-100 px-2 text-xs font-semibold leading-5 text-green-800"
            >
              View on Github
            </p>
          </div>
        </div>
      </div>
    </a>
  </li>,
  <li>
    <a
      className="block hover:bg-gray-50"
      href="https://github.com/gatsbyjs/gatsby/commit/4cf9ed6cd199683d62a0d78dea06436eddeb28a8"
    >
      <div
        className="px-4 py-4 sm:px-6"
      >
        <div
          className="flex items-center justify-between"
        >
          <p
            className="truncate text-sm font-medium text-indigo-600"
          >
            chore(docs): Release Notes for 4.24 (#36666)
          </p>
          <div
            className="ml-2 flex flex-shrink-0"
          >
            <p
              className="inline-flex rounded-full bg-green-100 px-2 text-xs font-semibold leading-5 text-green-800"
            >
              View on Github
            </p>
          </div>
        </div>
      </div>
    </a>
  </li>,
  <li>
    <a
      className="block hover:bg-gray-50"
      href="https://github.com/gatsbyjs/gatsby/commit/c2a3dfafa120e011a0742ccf40d114e6d3680d85"
    >
      <div
        className="px-4 py-4 sm:px-6"
      >
        <div
          className="flex items-center justify-between"
        >
          <p
            className="truncate text-sm font-medium text-indigo-600"
          >
            fix(gatsby): Don't throw when a path with special character is visite…
          </p>
          <div
            className="ml-2 flex flex-shrink-0"
          >
            <p
              className="inline-flex rounded-full bg-green-100 px-2 text-xs font-semibold leading-5 text-green-800"
            >
              View on Github
            </p>
          </div>
        </div>
      </div>
    </a>
  </li>,
]
`;
